lib/localisation/src/preprocessing/accelerometer.rs

```diff
@@ -1 +1,283 @@
+            upper_bound: q3 + bound_factor * (q3 - q1),
+        }
+    }
+}
+
+#[derive(Default)]
+pub struct AccelerometerPreprocessor {
+    num_reliable_accelerometers: i32,
+    reliable_accelerometers: [bool; K_NUM_ACCELEROMETERS],
+    num_outliers_per_accelerometer: [i32; K_NUM_ACCELEROMETERS],
+}
+
+impl AccelerometerPreprocessor {
+    pub fn new() -> Self {
+        Self {
+            num_reliable_accelerometers: K_NUM_ACCELEROMETERS as i32,
+            reliable_accelerometers: [true; K_NUM_ACCELEROMETERS],
+            num_outliers_per_accelerometer: [0; K_NUM_ACCELEROMETERS],
+        }
+    }
+
+    pub fn handle_outliers(
+        &mut self,
+        data: AccelerometerData<K_NUM_ACCELEROMETERS>,
+    ) -> Option<AccelerometerData<K_NUM_ACCELEROMETERS>> {
+        let quartiles = self.calculate_quartiles(&data)?;
+
+        let accelerometer_data = data
+            .iter()
+            .enumerate()
+            .map(|(i, &val)| {
+                if !self.reliable_accelerometers[i] {
+                    quartiles.q2
+                } else if val < quartiles.lower_bound || val > quartiles.upper_bound {
+                    self.num_outliers_per_accelerometer[i] += 1;
+                    quartiles.q2
+                } else {
+                    self.num_outliers_per_accelerometer[i] = 0;
+                    val
+                }
+            })
+            .collect();
+
+        Some(accelerometer_data)
+    }
+
+    pub fn calculate_quartiles(
+        &self,
+        data: &AccelerometerData<K_NUM_ACCELEROMETERS>,
+    ) -> Option<Quartiles> {
+        let quartiles: Quartiles;
+        if self.num_reliable_accelerometers == K_NUM_ACCELEROMETERS as i32 {
+            quartiles = self.get_quartiles(data);
+        } else if self.num_reliable_accelerometers == (K_NUM_ACCELEROMETERS as i32 - 1) {
+            const SIZE: usize = K_NUM_ACCELEROMETERS - 1;
+            let filtered_data: AccelerometerData<SIZE> = data
+                .iter()
+                .enumerate()
+                .filter(|(i, _)| self.reliable_accelerometers[*i])
+                .map(|(_, val)| *val)
+                .collect();
+            quartiles = self.get_quartiles(&filtered_data);
+        } else {
+            return None;
+        }
+
+        Some(quartiles)
+    }
+
+    pub fn process_data(
+        &mut self,
+        data: RawAccelerometerData<K_NUM_ACCELEROMETERS, K_NUM_AXIS>,
+    ) -> Option<AccelerometerData<K_NUM_ACCELEROMETERS>> {
+        let mut accelerometer_data: AccelerometerData<K_NUM_ACCELEROMETERS> =
+            AccelerometerData::from_slice(&[0.0; K_NUM_ACCELEROMETERS]).unwrap();
+        let mut magnitude: f32 = 0.0;
+
+        data.iter().enumerate().for_each(|(i, axis)| {
+            magnitude = 0.0;
+            axis.iter().for_each(|val| {
+                magnitude += val * val;
+            });
+            accelerometer_data[i] = magnitude.sqrt();
+        });
```

Comment:
```suggestion        
        let accelerometer_data: AccelerometerData<K_NUM_ACCELEROMETERS> = data
            .iter()
            .map(|axis| axis.iter().fold(0.0, |acc, val| acc + val * val).sqrt())
            .collect();
```
I think this is a nicer way to do this, avoids unnecessary `mut`s
Maybe i'm just too functional pilled tho